transfer-manager/contracts/TransferExecutor.sol

```diff
@@ -33,26 +35,41 @@ abstract contract TransferExecutor is Initializable, OwnableUpgradeable, ITransf
         LibAsset.Asset memory asset,
         address from,
         address to,
-        bytes4 transferDirection,
-        bytes4 transferType
+        address proxy
     ) internal override {
-        if (asset.assetType.assetClass == LibAsset.ETH_ASSET_CLASS) {
-            to.transferEth(asset.value);
-        } else if (asset.assetType.assetClass == LibAsset.ERC20_ASSET_CLASS) {
-            (address token) = abi.decode(asset.assetType.data, (address));
-            IERC20TransferProxy(proxies[LibAsset.ERC20_ASSET_CLASS]).erc20safeTransferFrom(IERC20Upgradeable(token), from, to, asset.value);
-        } else if (asset.assetType.assetClass == LibAsset.ERC721_ASSET_CLASS) {
+        if (asset.assetType.assetClass == LibAsset.ERC721_ASSET_CLASS) {
+            //not using transfer proxy when transfering from this contract
             (address token, uint tokenId) = abi.decode(asset.assetType.data, (address, uint256));
             require(asset.value == 1, "erc721 value error");
-            INftTransferProxy(proxies[LibAsset.ERC721_ASSET_CLASS]).erc721safeTransferFrom(IERC721Upgradeable(token), from, to, tokenId);
+            if (from == address(this)){
+                IERC721Upgradeable(token).safeTransferFrom(address(this), to, tokenId);
+            } else {
+                INftTransferProxy(proxy).erc721safeTransferFrom(IERC721Upgradeable(token), from, to, tokenId);
+            }
+        } else if (asset.assetType.assetClass == LibAsset.ERC20_ASSET_CLASS) {
+            //not using transfer proxy when transfering from this contract
+            (address token) = abi.decode(asset.assetType.data, (address));
+            if (from == address(this)){
+                IERC20Upgradeable(token).transfer(to, asset.value);
+            } else {
+                IERC20TransferProxy(proxy).erc20safeTransferFrom(IERC20Upgradeable(token), from, to, asset.value);
+            }
         } else if (asset.assetType.assetClass == LibAsset.ERC1155_ASSET_CLASS) {
+            //not using transfer proxy when transfering from this contract
             (address token, uint tokenId) = abi.decode(asset.assetType.data, (address, uint256));
-            INftTransferProxy(proxies[LibAsset.ERC1155_ASSET_CLASS]).erc1155safeTransferFrom(IERC1155Upgradeable(token), from, to, tokenId, asset.value, "");
+            if (from == address(this)){
+                IERC1155Upgradeable(token).safeTransferFrom(address(this), to, tokenId, asset.value, "");
+            } else {
+                INftTransferProxy(proxy).erc1155safeTransferFrom(IERC1155Upgradeable(token), from, to, tokenId, asset.value, "");  
+            }
+        } else if (asset.assetType.assetClass == LibAsset.ETH_ASSET_CLASS) {
+            if (to != address(this)) {
```

Comment:
is it better to check if from is correct here? Just to double-check. from can be only this?